docs: Update Clef/Rules page (#25335)
ethereum/go-ethereum: docs/_clef/Rules.md

---
title: Rules
sort_key: C
---

This page provides a fairly low-level explanation for how rules are implemented in 
Clef. It is a good idea to read the [Introduction to Clef](/docs/clef/introduction)
and the [Clef tutorial](/docs/clef/tutorial) before diving in to this page.

{:toc}
-   this will be removed by the toc
  
## Introduction

Rules in Clef are sets of conditions that determine whether a given action can be
signed automatically without requiring manual approval from the user. This can be
useful for automatically approving transactions between a user's own accounts, or
to approve patterns that are commonly used by applications. This is a UI improvement
for users making lots of transactions and also a security improvement because transactions
that do not fit the pattern defined by a ruleset will not be signed unless they are 

> reviewed and manually approved by the end user.

Reviewer: Rules only govern "approve" vs " disapprove" vs "let someone else decide". The password required for signing is not part of the rules, and the feature to allow clef to remember passwords is independent. 

Also: `and also a security improvement` -- not sure what you mean. Security improvement over what? Over manual review? Hardly, I'd say.

Author: Fair points, thanks. I think I meant 'security improvement' compared to having an unlocked account in Geth but on reflection that doesn't make so much sense.

```suggestion
Rules in Clef are sets of conditions that determine whether a given action can be
approved automatically without requiring manual intervention from the user. This can be
useful for automatically approving transactions between a user's own accounts, or
approving patterns that are commonly used by applications. Automatic signing also 
requires Clef to have access to account passwords which is configured independently 
of the ruleset. 
```

Rules can define arbitrary conditions such as:

* Auto-approve 10 transactions with contract `CasinoDapp`, with value between `0.05 ether` and 
  `1 ether` per 24h period.

* Auto-approve transactions to contract `Uniswapv2` with `value` up to 1 ether, if 
  `gas < 44k` and `gasPrice < 40Gwei`.

* Auto-approve signing if the data to be signed contains the string `"approve_me"`.

* Auto-approve any requests to list accounts in keystore if the request arrives over IPC

Because the rules are Javascript files they can be customized to implement any arbitrary logic on
the available request data.

This page will explain how rules are implemented in Clef and how best to manage credentials
when automatic rulesets are enabled.

## Rule Implementation

> Although rules in Clef are written in Javascript, Clef itself is written in Go. To

Author: ```suggestion
```

implement the Javascript rules in Clef, a virtual machine 
([OttoVM](https://github.com/robertkrimen/otto)) is embedded into the `signer` binary
that parses and interprets the Javascript code and imposes their logic in Clef. This 

> virtual machine is called the 'ruleset engine'.

Reviewer: It's not actualy OttoVM any longer. On the whole, I'm sceptical as to whether this paragraph is warranted. Too low level IMO

Author: ```suggestion
```

| Anything else             | Pass decision to UI for manual approval   |
 
 
There are some additional noteworthy implementation details that are important 
for defining rules correctly in `ruleset.js`:

* The code in `ruleset.js` **cannot** load external Javascript files. 
  
* The Javascript engine can access `storage` and `console`
  
* The only preloaded library in the Javascript environment is 
  `bignumber.js` version `2.0.3`.
  
* Each invocation is made in a fresh virtual machine meaning data cannot be 
  stored in global variables between invocations.
  
* Since no global variable storage is available, disk backed `storage` must be 
  used - rules should not rely on ephemeral data.
  
* Javascript API parameters are always objects. This ensures parameters are 
  accessed by _key_ to avoid misordering errors. 
  
* Otto VM uses ES5. ES6-specific features (such as Typed Arrays) are not supported.
  
* The regular expression engine (re2/regexp) in Otto VM is not fully compatible 
  with the [ECMA5 specification](https://tc39.es/ecma262/#sec-intro).
  
* [Strict mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode) 

>   is not supported. "Use strict" will parse but it does nothing.

Reviewer: ```suggestion
  is not supported. "use strict" will parse but it does nothing.
```

}
```

When a request is made via the external API, the logic flow is as follows:

* Request is made to the `signer` binary using external API
   
* `signer` calls the UI - in this case the ruleset engine

* UI evaluates whether the call conforms to rules in an attested rulefile

* Assuming the call returns "Approve", request is signed.

There are five possible outcomes from the ruleset engine that are 
handled in different ways:
 
| Return value      |    Action                                 |
| ------------------| ----------------------------------------- |
| Approve           | Auto-approve request       			    |
| Reject            | Auto-reject request        				|
| Error             | Pass decision to UI for manual approval   |
| Unexpected value  | Pass decision to UI for manual approval   |

> | Nothing           | Pass decision to UI for manual approval   |

Reviewer: I'd say that there are three outcomes. 

1. The literal string `Approve`. 
2. The literal string `Reject`. 
3. Anything else

Author: ```suggestion
 
| Return value      |    Action                                 |
| ------------------| ----------------------------------------- |
| "Approve"           | Auto-approve request       			    |
| "Reject"            | Auto-reject request        				|
| Anything else             | Pass decision to UI for manual approval   |
```

the necessary credentials, i.e. account passwords, to decrypt keyfiles. 
These are stored encrypted as follows:

When the `signer` is started it generates a seed that is locked with a 
user specified password. The seed is saved to a location that defaults to
`$HOME/.clef/masterseed.json`. The `seed` itself is a blob of bytes.

The `signer` uses the `seed` to:

* Generate the `path` where the configuration and credentials data are stored.
  * `$HOME/.clef/790046d38025/config.json`
  * `$HOME/.clef/790046d38025/credentials.json`
* Generate the encryption password for the config and credentials files.

`config.json` stores the hashes of any attested rulesets. `credentials.json`
stores encrypted account passwords. The masterseed is required to decrypt
these files. The decrypted account passwords can then be used to decrypt keyfiles.

## Security

### The Javascript VM
The downside of the very flexible rule implementation included in Clef is
that the `signer` binary needs to contain a Javascript engine. This is an
additional attack surface. The only viable attack is for an adversary to 
somehow extract cryptographic keys from memory during the Javascript VM execution.

> The hash-based rule attestation condition means the actual Javascript code 
> executed by the Javascript engine is not a viable attack surface. A much simpler
> attack would be to surreptitiously insert an attested "always-approve" rule 
> instead of attempting to exploit the Javascript virtual machine. The Javascript

Reviewer: ```suggestion
The hash-based rule attestation condition means the actual Javascript code 
executed by the Javascript engine is not a viable attack surface -- since if the attacker can control the ruleset, a much simpler
attack would be to surreptitiously insert an attested "always-approve" rule 
instead of attempting to exploit the Javascript virtual machine. The Javascript
```

vulnerabilities, not have there been any security problems identified for the
similar Javascript VM implemented in Geth.

### Writing rules

Since the user has complete freedom to write custom rules, it is plausible that those rules
could create unintended security vulnerabilities. This can only really be protected by 
coding very carefully and trying to test rulesets (e.g. on a private testnet) before 
implementing them on a public network.

Javascript is very flexible but also easy to write incorrectly. For example, users
might assume that javascript can handle large integers natively rather than explicitly
using `bigInt`. This is an error commonly encountered in the Ethereum context when 
users attempt to multiply `gas` by `gasCost`.

It’s unclear whether any other language would be more secure - there is alwas the possibility
of implementing an insecure rule.

### File security

A ruleset should have its file permissions set to prevent attacks where an adversary
drops files onto the user's disk. Specifically, `ruleset.js` should only be loaded if it
is set to `readonly` (`r-xx-xx-x`). To update the ruleset, it should be explicitly made

> writeable then set back to `readonly` before being reloaded.

Reviewer: Please delete this. The original doc included this as "potential todos", but then instead we went with the "attestation" solution. So the attack of replacing the file on disk is moot.

Author: ```suggestion
```
